### PR TITLE
[BEAM-4681] Add support for portable timers in Flink streaming mode 

### PR DESCRIPTION
**Pull request by mxm**
This adds support for portable timers to the Flink Runner in streaming mode. Batch support will be added in a follow-up.

The `UsesTimersInParDo` tests of Java ValidatesPortableRunner have been run on it and they are passing. We can't enable them yet because they currently run only for batch. See https://issues.apache.org/jira/browse/BEAM-6009.

To enable the Python ValidatesPortableRunner tests https://issues.apache.org/jira/browse/BEAM-5999 needs to be addressed.

CC @tweise @angoenka @robertwb 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 1]
```diff
@@ -359,16 +543,82 @@ public void finishBundle() {
...
+                "Received a null value from the SDK harness for %s",
+                inputCollectionId);
+        if (tag != null) {
+          // process regular elements
+          outputManager.output(tag, windowedValue);
+        } else {
+          // process timer elements
+          // TODO This is ugly. There should be an easier way to retrieve the
+          String timerPCollectionId =
+              inputCollectionId.substring(0, inputCollectionId.length() - ".out:0".length());
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: @lukecwik I couldn't figure out how to retrieve the correct name for the collection of the timer input. I believe this needs to be fixed.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -242,6 +242,7 @@ class BeamModulePlugin implements Plugin<Project> {
       excludeCategories 'org.apache.beam.sdk.testing.UsesMapState'
       excludeCategories 'org.apache.beam.sdk.testing.UsesSetState'
       excludeCategories 'org.apache.beam.sdk.testing.UsesTestStream'
+      // TODO Enable test once timer-support for batch is merged
```
Comment (tweise) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: It would be really nice to get this configuration out of this file as it belongs to the Java SDK. Worth filing a JIRA for follow-up?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 3]
```diff
@@ -442,22 +443,21 @@ public void close() throws Exception {
     }
   }
 
-  private long getPushbackWatermarkHold() {
+  long getPushbackWatermarkHold() {
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: protected?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 4]
```diff
@@ -669,7 +669,7 @@ private void checkInvokeFinishBundleByTime() {
     }
   }
 
-  private void invokeFinishBundle() {
+  void invokeFinishBundle() {
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: protected, since the intention is to make it accessible to subclasses?

File: runners/flink/src/test/java/org/apache/beam/runners/flink/PortableTimersExecutionTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * Tests the state and timer integration of {@link
+ * org.apache.beam.runners.flink.translation.wrappers.streaming.ExecutableStageDoFnOperator}.
+ */
+@RunWith(Parameterized.class)
+public class PortableTimersExecutionTest implements Serializable {
```
Comment (tweise) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/PortableTimersExecutionTest.java]: Due to the more heavyweight nature of these tests, I was wondering if state and timers could be covered by a single test? (outside of this PR)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 6]
```diff
@@ -2839,19 +2841,27 @@ public void testAbsoluteProcessingTimeTimerRejected() throws Exception {
 
             @ProcessElement
             public void processElement(@TimerId(timerId) Timer timer) {
-              timer.set(new Instant(0));
+              try {
+                timer.set(new Instant(0));
+                fail("Should have failed due to processing time with absolute timer.");
```
Comment (tweise) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: The error message is a b it confusing. Should it refer to "invalid processing time"?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 7]
```diff
@@ -152,7 +163,8 @@ public void open() throws Exception {
     stageBundleFactory = stageContext.getStageBundleFactory(executableStage);
     stateRequestHandler = getStateRequestHandler(executableStage);
     progressHandler = BundleProgressHandler.unsupported();
-    outputQueue = new LinkedBlockingQueue<>();
+
+    super.open();
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: add a comment why it has to be in this order?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 8]
```diff
@@ -219,31 +231,43 @@ private BagUserStateFactory(
       return new StateRequestHandlers.BagUserStateHandler<K, V, W>() {
         @Override
         public Iterable<V> get(K key, W window) {
-          prepareStateBackend(key, keyCoder);
-          StateNamespace namespace = StateNamespaces.window(windowCoder, window);
-          BagState<V> bagState =
-              stateInternals.state(namespace, StateTags.bag(userStateId, valueCoder));
-          return bagState.read();
+          synchronized (keyedStateBackend) {
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: The SDK harness issues concurrent state access for different keys?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 9]
```diff
@@ -270,8 +294,22 @@ public void setKeyContextElement1(StreamRecord record) throws Exception {
 
   @Override
   public void setCurrentKey(Object key) {
-    throw new UnsupportedOperationException(
-        "Current key for state backend can only be set by state requests from SDK workers.");
+    if (stateful && usesTimers) {
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: I think it would be more readable to have an outer `if (usesTimers)` and then a nested `if (stateful)` - but how does synchronizing just `setCurrentKey` help, how is the current key then used (not synchronized?) when timers are present?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 10]
```diff
@@ -304,16 +342,141 @@ protected void addSideInputValue(StreamRecord<RawUnionValue> streamRecord) {
...
+                throw new RuntimeException(
+                    "Failed to process pushed back watermark after finished bundle.", e);
+              }
+            });
+      }
+    }
+    super.processWatermark(mark);
+  }
+
+  private static class SdkHarnessDoFnRunner<InputT, OutputT>
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: What's the reason to change it to static (adds all the constructor clutter)?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 11]
```diff
@@ -304,16 +342,141 @@ protected void addSideInputValue(StreamRecord<RawUnionValue> streamRecord) {
...
+    //    have been emitted to the output stream.
+    //
+    // 2) Put a hold on the output watermark for as long as the current
+    //    bundle has not been finished. We have to remember to manually
+    //    finish the bundle in case we receive the final watermark.
+    //    To avoid latency, we should process this watermark again as
+    //    soon as the current bundle is finished.
+    //
+    // Approach 1) is the easiest, yet 2) gives better throughput due
+    // to the bundle getting cut on every watermark. So we have
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: "the bundle not getting cut"

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 9 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Due to the concurrent access the synchronization is actually quite hard. I think I have pushed a good solution now.

**Comment by mxm** [comment]
Rebased

**Comment by mxm** [comment]
Run Website PreCommit

**Comment by mxm** [comment]
Run Website PreCommit

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java (hunk 3 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: The reason it changed from private is to make it available to the subclass. IMO (like other similar modifier changes) it should be `protected`, because we don't want to force subclasses to the same package. `final` is orthogonal.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 12]
```diff
@@ -442,22 +443,21 @@ public void close() throws Exception {
     }
   }
 
-  private long getPushbackWatermarkHold() {
+  long getPushbackWatermarkHold() {
     return pushedBackWatermark;
   }
 
+  void setPushedBackWatermark(long watermark) {
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: protected

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 7 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Even with this explanation it isn't obvious :) Please add a comment.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 13]
```diff
@@ -304,16 +344,126 @@ protected void addSideInputValue(StreamRecord<RawUnionValue> streamRecord) {
...
+    //    received by the outputQueue (see below), where they will
+    //    have been emitted to the output stream.
+    //
+    // 2) Put a hold on the output watermark for as long as the current
+    //    bundle has not been finished. We have to remember to manually
+    //    finish the bundle in case we receive the final watermark.
+    //    To avoid latency, we should process this watermark again as
+    //    soon as the current bundle is finished.
+    //
+    // Approach 1) is the easiest, yet 2) gives better throughput due
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: It is a latency throughput tradeoff. But we have bundle max millis to control latency.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 10 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Yeah, you can probably take the argument either way. I see this in few other places in the Flink runner and I see your point regarding the required dependencies. On the other hand these redundant constructor argument lists and member declarations reduce readability, sometimes make it hard to find what really matters and overall provide little benefit when the class is an internal of the outer class and fully encapsulated. Not a blocker for me though.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 1 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: This looks ugly indeed.. Is this going to go away with the changes @robertwb mentioned last week?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 14]
```diff
@@ -359,16 +534,86 @@ public void finishBundle() {
...
+              StateNamespace namespace =
+                  StateNamespaces.window(windowCoder, (BoundedWindow) window);
+              TimerSpec timerSpec = extractTimerSpec(timerReference);
+              TimerInternals.TimerData timerData =
+                  TimerInternals.TimerData.of(
+                      timerPCollectionId,
+                      namespace,
+                      timer.getTimestamp(),
+                      timerSpec.getTimeDomain());
+              timerDataConsumer.accept(windowedValue, timerData);
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: This is another example of unnecessary code complexity. Had to navigate a chain in the IDE when this otherwise could just be a simple method call on `ExecutableStageDoFnOperator`.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/ExecutableStageDoFnOperatorTest.java [hunk 15]
```diff
@@ -225,18 +229,22 @@ public String getId() {
...
-                    "pCollectionId",
+                    "input",
                     input -> {
                       /* Ignore input*/
                     });
               }
 
               @Override
               public void close() throws Exception {
+                if (onceEmitted) {
```
Comment (tweise) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/ExecutableStageDoFnOperatorTest.java]: Why is this required? Duplicate call to `close()`?

**Comment by tweise** [comment]
Exciting to see this coming along! Functionality wise it looks great, just a few code style related comments remaining.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/ExecutableStageDoFnOperatorTest.java (hunk 15 above)
Comment (mxm) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/ExecutableStageDoFnOperatorTest.java]: Close is only called once per bundle but there are two bundles now due to the watermark processing.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 14 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: I agree it can be confusing. On the other hand, this decouples the timer generation code and the timer consumer code.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 13 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Clarifying this.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java (hunk 12 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: will change

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 6 above)
Comment (mxm) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Yes, but not in processing time: https://github.com/apache/beam/commit/86bcf26c965f876be2e67aea769de3406ef25a97#diff-95b3ede99105b81444964de712a52386R967

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 10 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: We could move this out of the operator now which would increase readability. If we don't split it into a separate static class, we wouldn't be able to do that. I like encapsulation and ExecutableStageDoFnOperator doesn't encourage clean design with its heavy use of virtual methods. I think we have to do a separate refactoring to further increase readability.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 1 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: There is #6986 but this just prepares the Java code for handling the Python representation.

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Found a way, for example:

```java
stageBundleFactory
.getProcessBundleDescriptor()
// yes this is required twice
.getProcessBundleDescriptor()
.getPcollectionsMap()
.get("ParDo(Anonymous)2/ParMultiDo(Anonymous).foo.out:0")
.getUniqueName()
```
=> `"ParDo(Anonymous)2/ParMultiDo(Anonymous).foo"`

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner

**Comment by mxm** [comment]
I've addressed your comments @tweise.

**Comment by tweise** [comment]
Run Java Flink PortableValidatesRunner

**Comment by tweise** [comment]
Run Python Flink ValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 10 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: I would agree with that if `SdkHarnessDoFnRunner` was a component with a prospect of being used independently. Instead it is an implementation detail of the operator that is tightly coupled. Ideally the class wouldn't even exist. I would probably go the opposite way and just delegate all methods to the operator, which would remove the clutter can keep the class super small. It could even be an anonymous class.  Again, this is just for sake of discussion and not actionable for this PR.

Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Regarding the operator as a whole and as discussed in the past, constraints are mostly inherited and would need to be addressed by refactoring the base. I'm not sure about the timing for this though, if it should wait until legacy runner goes away (there is still significant work for portable runner to become adequate replacement for Java pipelines).

**Comment by mxm** [comment]
Looks like a Jenkins problem, looking into it.

```
18:55:59 Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x000000019a300000, 1915748352, 0) failed; error='Cannot allocate memory' (errno=12)
18:55:59 #
18:55:59 # There is insufficient memory for the Java Runtime Environment to continue.
18:55:59 # Native memory allocation (mmap) failed to map 1915748352 bytes for committing reserved memory.```

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 16]
```diff
@@ -359,13 +533,100 @@ public void finishBundle() {
...
+      }
+    }
+
+    private String extractTimerPCollectionId(String outputPCollectionId) {
+      return stageBundleFactory
+          .getProcessBundleDescriptor()
+          .getProcessBundleDescriptor()
+          .getPcollectionsMap()
+          .get(outputPCollectionId)
+          .getUniqueName();
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: The name here is somehow not stable anymore after #6986.

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: @robertwb Reverted #6986 and it works again.

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner

**Comment by mxm** [comment]
Run Python Flink ValidatesRunner

**Comment by mxm** [comment]
Run Website PreCommit

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 16 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: For example, I have `ParDo.timer.foo.out:0` which resolves to `ParDo.timer.output` instead of `ParDo.timer.foo`.

**Comment by tweise** [comment]
Run Python Flink ValidatesRunner

**Comment by tweise** [comment]
Run Python Flink ValidatesRunner

**Comment by mxm** [comment]
Run Python Flink ValidatesRunner

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 16 above)
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Unless I'm mistaken, extractTimerPCollectionId should not be needed at all anymore, as timerReferenceMap is indexed by outputPCollectionId itself.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java [hunk 17]
```diff
@@ -359,13 +534,100 @@ public void finishBundle() {
...
+              Preconditions.checkNotNull(
+                  (Timer) ((KV) windowedValue.getValue()).getValue(),
+                  "Received null Timer from SDK harness: %s",
+                  windowedValue);
+          LOG.debug("Timer received: {} {}", outputPCollectionId, timer);
+          for (Object window : windowedValue.getWindows()) {
+            StateNamespace namespace = StateNamespaces.window(windowCoder, (BoundedWindow) window);
+            TimerInternals.TimerData timerData =
+                TimerInternals.TimerData.of(
+                    timerCollectionId, namespace, timer.getTimestamp(), timerSpec.getTimeDomain());
```
Comment (robertwb) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: The notion "timerCollectionId" is ambiguous, as timers are associated with two distinct PCollections at execution time. 

Timers are absolutely determined by their transform name and local name; probably it makes sense to use that (let's call it fullyQualifiedTimerName) for the TimerData here, and then also have a mapping fullyQualifiedTimerName -> timerInputPCollectionId for plumbing timers to the SDK, and a map timerOutputPCollectionId -> fullyQualifiedTimerName for recording timers from the SDK (like we do here).

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 16 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: It is still necessary. The timerReferenceMap gives me `ParDo.timer.foo` but while processing the timer elements, I get `ParDo.timer.foo.out:0`. The issue seems to be that the output collection ids are not set correctly.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 17 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: >Timers are absolutely determined by their transform name and local name; probably it makes sense to use that

That's what is used here. The keys of `timerReferenceMap` are transform name + local name. The "extractTimerPCollectionId" method does the mapping of output collection to timer name.

>and then also have a mapping fullyQualifiedTimerName -> timerInputPCollectionId for plumbing timers to the SDK, and a map timerOutputPCollectionId -> fullyQualifiedTimerName for recording timers from the SDK (like we do here).

I couldn't find a nice way to create a map timerOutputPCollectionId -> fullyQualifiedTimerName.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 16 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: >Unless I'm mistaken, extractTimerPCollectionId should not be needed at all anymore, as timerReferenceMap is indexed by outputPCollectionId itself.

It is still necessary. The timerReferenceMap gives me `ParDo.timer.foo` for the output name of timer "foo", when it is `ParDo.timer.foo.out:0` at runtime.  I believe there lies the problem.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 17 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: I think it will be best we merge this PR and revert #6986.  BEAM-5999 can then rebase and address this issue.

**Comment by tweise** [comment]
Run Java Flink PortableValidatesRunner

**Comment by mxm** [comment]
PVR are extremely slow because the Docker environment takes long to come up for every test:

```
INFO org.apache.beam.runners.fnexecution.environment.DockerEnvironmentFactory - Still waiting for startup of environment jenkins-docker-apache.bintray.io/beam/java for worker id 1
```

Eventually the 100 minute timeout kills the job.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 17 above)
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: OK, the problem is that there are multiple copies of components and pcollections floating around...

The right way to solve this is to patch in https://github.com/mxm/beam/pull/1/commits/55baf15befea582a3fc7796f9c0a4a031c3485b3 and then use stageBundleFactory.getProcessBundleDescriptor().getTimerSpecs() in SdkHarnessDoFnRunner to construct timerReferenceMap or whatever else is needed, which has the full information.

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java (hunk 17 above)
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: https://github.com/mxm/beam/pull/1 This also avoids hacks to guess the output name from the input name.

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java]: Neat. With your fix I was able to build a `timerOutputIdToSpecMap` from output collection id to TimerSpec. This also got rid of some boiler plate code which I previously used to extract the TimerSpec.

**Comment by mxm** [comment]
Run Python Flink ValidatesRunner

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner
